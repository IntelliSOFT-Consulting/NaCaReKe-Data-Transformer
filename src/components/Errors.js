/* eslint-disable react/button-has-type */
/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import PropTypes from 'prop-types';
import { Table, Tag, Popconfirm } from 'antd';
import { closestMatch } from 'closest-match';
import correctAddress from '../corrections/address';
import subCounties from '../data/scounties';

export default function Errors({ errors, setErrors, data, setData }) {
  const columns = [
    {
      title: 'Sub Counties',
      key: 'addr',
      dataIndex: 'addr',
      render: (tags) => (
        tags?.map((tag) => {
          const alt = closestMatch(tag, subCounties);
          const handleCorrect = async (sc, replace) => {
            const newData = await correctAddress(data, sc, replace);
            await setData(newData);
            setErrors([{ ...errors[0], addr: errors[0].addr.filter((t) => t !== tag) }]);
          };

          return (
            <>
              {alt ? (
                <Popconfirm
                  title={(
                    <p>
                      Replace with
                      {' '}
                      <strong>{alt}</strong>
                      ?
                    </p>
)}
                  onConfirm={() => handleCorrect(tag, alt)}
                  okText="Yes"
                  cancelText="No"
                >
                  <Tag color="gold" key={tag}>
                    {tag}
                  </Tag>
                </Popconfirm>
              ) : (
                <Tag color="volcano" key={tag}>
                  {tag}
                </Tag>
              )}
            </>
          );
        })
      ),
    },
    {
      title: 'TOP',
      dataIndex: 'top',
      key: 'top',
      render: (tops) => (
        <p>{tops?.join(', ')}</p>
      ),
    },
    {
      title: 'MOR',
      dataIndex: 'mor',
      key: 'mor',
      render: (mors) => (
        <p>{mors?.join(', ')}</p>
      ),
    },

  ];

  return (
    <div>
      {errors && Object.keys(errors).length > 0 && (
      <a
        href={`data:text/json;charset=utf-8,${encodeURIComponent(
          JSON.stringify(errors),
        )}`}
        download="errors.json"
        className="err-container"
      >
        Download Json
      </a>
      )}
      <Table columns={columns} dataSource={errors} />
    </div>
  );
}

Errors.propTypes = {
  errors: PropTypes.array.isRequired,
  setErrors: PropTypes.func.isRequired,
  data: PropTypes.array.isRequired,
  setData: PropTypes.func.isRequired,
};
