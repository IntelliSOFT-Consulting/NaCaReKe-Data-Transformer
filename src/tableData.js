/* eslint-disable import/no-mutable-exports */
let data = [
  [
    'RECS',
    'RECS (desc)',
    'CHEC',
    'CHEC (desc)',
    'AGE',
    'ADDR',
    'ADDR (cat)',
    'ADDR (desc)',
    'INCID',
    'TOP',
    'TOP (cat)',
    'TOP (desc)',
    'LTERALITY',
    'LTERALITY (desc)',
    'MOR',
    'MOR (desc)',
    'BEH',
    'BEH (desc)',
    'BAS',
    'BAS (desc)',
    'MPCODE',
    'MPSEQ',
    'MPTOT',
    'I10',
    'UPDATE',
    'ICCC',
    'OBSOLETEFLAGTUMOURTABLE',
    'TUMOURID',
    'PATIENTIDTUMOURTABLE',
    'PATIENTRECORDIDTUMOURTABLE',
    'TUMOURUPDATEDBY',
    'TUMOURUNDUPLICATIONSTATUS',
    'LABNUMBER',
    'LABREPORTAVAILABLEPTFILE',
    'LABREPORTAVAILABLEPTFILE (desc)',
    'CD4COUNT',
    'STATUSINDICATEDONCLINICALNOTES',
    'STATUSINDICATEDONCLINICALNOTES (desc)',
    'SPECIFICALLYNEGATIVE',
    'SPECIFICALLYNEGATIVE (desc)',
    'SPECIFICALLYPOSITIVE',
    'SPECIFICALLYPOSITIVE (desc)',
    'OTHERCONCURRENTILLNESS',
    'OTHERADDRESS',
    'GRADE',
    'GRADE (desc)',
    'STAGE',
    'STAGE (desc)',
    'REGNO',
    'PERS',
    'PERS (desc)',
    'FIRSTN',
    'MIDDLNAME',
    'FAMN',
    'SEX',
    'SEX (desc)',
    'BIRTHD',
    'TRIB',
    'TRIB (desc)',
    'OCCU',
    'OCCU (desc)',
    'DLC',
    'STAT',
    'STAT (desc)',
    'MIDN',
    'OBSOLETEFLAGPATIENTTABLE',
    'PATIENTRECORDID',
    'PATIENTUPDATEDBY',
    'PATIENTUPDATEDATE',
    'PATIENTRECORDSTATUS',
    'PATIENTCHECKSTATUS',
    'SURTCT',
    'SURTCT (desc)',
    'SURGERYDATE',
    'CHEMOTREAT',
    'CHEMOTREAT (desc)',
    'CHEMOTREATMENTDATE',
    'IMMUNOTREAT',
    'IMMUNOTREAT (desc)',
    'IMMUNOTREATMENTDATE',
    'HORMONETREATMENT',
    'HORMONETREATMENT (desc)',
    'HORMONETHERAPY',
    'RADIOTREAT',
    'RADIOTREAT (desc)',
    'RADIOTREATMENTDATE',
    'OTHERTREATMENT',
    'OTHERDATE',
    'HOSPITALNUMBER',
    'DEATHCAUSE',
    'REMARKSIFANY',
    'IDENTIFICATIONNO',
    'MARITALSTATUS',
    'MARITALSTATUS (desc)',
    'TELEPHONENO',
    'TELEPHONENUMBERNOK',
    'RELIGION',
    'RELIGION (desc)',
    'EDUCATIONLEVEL',
    'EDUCATIONLEVEL (desc)',
    'SMOKINGSTATUS',
    'SMOKINGSTATUS (desc)',
    'DRINKINGSTATUS',
    'DRINKINGSTATUS (desc)',
  ],
  [
    1,
    'Confirmed',
    1,
    'Done: OK',
    1,
    101,
    'Nairobi',
    'Westlands',
    2018010,
    250,
    'PANCREAS',
    'Head of pancreas',
    9,
    'Unknown',
    8140,
    'Adenocarcinoma, NOS',
    3,
    'Malignant',
    7,
    'Histology of primary',
    null,
    0,
    1,
    'C250',
    '25/11/2019',
    '11f',
    0,
    'KH0150890101',
    'KH015089',
    'KB43743847',
    'carolyne',
    null,
    null,
    9,
    'Unknown',
    null,
    9,
    'Unknown',
    9,
    'Unknown',
    9,
    'Unknown',
    null,
    null,
    9,
    'Unknown',
    9,
    'Unknown',
    'KH015089',
    null,
    'Invalid code.',
    'John',
    'Doe',
    'Jane',
    1,
    'Male',
    '99/99/2016',
    2,
    'Kangemi',
    8,
    'Child',
    '30/11/2018',
    1,
    'Alive',
    'John',
    null,
    'KB43743847',
    'Jean',
    '25/11/2018',
    0,
    null,
    9,
    'Unknown',
    null,
    2,
    'Yes',
    null,
    1,
    'Unknown',
    null,
    9,
    'Unknown',
    null,
    9,
    'Unknown',
    null,
    null,
    null,
    3293232,
    null,
    null,
    null,
    5,
    'Child',
    '7843434343',
    '73223928382 Michael',
    1,
    'Muslim',
    7,
    'Child',
    9,
    'Unknown',
    9,
    'Unknown',
  ],
];

const positions = {
  lefts: [
    'CHEMOTREAT (true)',
    'IMMUNOTREAT (true)',
    'HORMONETREATMENT (true)',
    'RADIOTREAT (true)',
    'SURGERY (true)',
    'HIV(true)',
    'SPECIFICALLYPOSITIVE (true)',
    'OTHERTREATMENT(TRUE)',
    'BAS (true)',
    'TELEPHONENO(+254)',
    'OTHERCONCURRENTILLNESS(true)',
  ],
  rights: [
    'MOR(Matching NCI Codes)',
    'TOP(Matching NCI Codes)',
    'ADDR (Sub County)',
    'WARD',
    'NOKNUMBER',
    'NOKNAME',
    'ADDR (County)',
  ],
  yes: [
    'CHEMOTREAT (true)',
    'IMMUNOTREAT (true)',
    'HORMONETREATMENT (true)',
    'RADIOTREAT (true)',
    'SURGERY (true)',
  ],
};

let visible = true;

const setVisible = () => {
  visible = !visible;
};

const setData = (datas) => {
  data = datas;
};

export { setData, data, visible, setVisible, positions };
