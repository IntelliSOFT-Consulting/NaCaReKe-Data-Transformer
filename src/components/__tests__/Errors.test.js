/* eslint-disable no-undef */
import { data } from '../../tableData';
import Errors from '../Errors';

it('renders correctly', () => {
  const spy = sinon.spy();
  const wrapper = shallow(
    <Errors
      errors={[]}
      setErrors={spy}
      data={data}
      setData={spy}
    />,
  );
  expect(wrapper).toMatchSnapshot();
});
