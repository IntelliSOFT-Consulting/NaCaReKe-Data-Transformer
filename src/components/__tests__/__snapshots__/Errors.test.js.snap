// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <Table
    columns={
      Array [
        Object {
          "dataIndex": "addr",
          "key": "addr",
          "render": [Function],
          "title": "Sub Counties",
        },
        Object {
          "dataIndex": "top",
          "key": "top",
          "render": [Function],
          "title": "TOP",
        },
        Object {
          "dataIndex": "mor",
          "key": "mor",
          "render": [Function],
          "title": "MOR",
        },
      ]
    }
    dataSource={Array []}
    rowKey="key"
  />
</div>
`;
